src/hmem_cuda.c

```diff
@@ -482,6 +494,40 @@ static int cuda_hmem_verify_devices(void)
 	return FI_SUCCESS;
 }
 
+/*
+ * Loop over pairs of available devices to deduce the overall P2P capabilities
+ */
+static int cuda_hmem_detect_p2p_access_support(void)
+{
+	CUresult cuda_ret;
+	CUdevice dev, peer;
+	int can_access_peer = 1;
+
+	if (cuda_attr.device_count <= 1)
+		return FI_SUCCESS;
+
+	for (dev = 0, peer = 1; peer < cuda_attr.device_count; ++dev, ++peer) {
+		cuda_ret = ofi_cuDeviceCanAccessPeer(&can_access_peer, dev, peer);
+		if (CUDA_SUCCESS != cuda_ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"Failed to detect support for peer-to-peer "
+				"access between CUDA devices via "
+				"cuDeviceCanAccessPeer(): %s:%s\n",
+				ofi_cudaGetErrorName(cuda_ret),
+				ofi_cudaGetErrorString(cuda_ret));
+			return -FI_EIO;
+		}
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Peer access from CUDA device %d -> CUDA device %d : %s\n",
+			dev, peer, can_access_peer ? "Yes" : "No");
+		if (!can_access_peer)
+			return FI_SUCCESS;
```

Comment:
So if can_access_peer is false, we still return FI_SUCCESS? what is the expected return code for this function. Can you clarify it in function's docstring?

Reply:
Ack, added a Doxygen comment for this function

Reply:
Could the doc string mention that p2p_access_supported  will be set on FI_SUCCESS?

Reply:
> Could the doc string mention that p2p_access_supported will be set on FI_SUCCESS?

That's only true in the case where the `for` loop finishes (i.e. p2p access available across all visible devices), but I can document this.